In [3]:
import spacy
from spacy import displacy
from IPython.display import HTML, display

# spacy.cli.download("en_core_web_sm")
nlp = spacy.load("en_core_web_sm")

text = "I saw the man with a telescope."

doc = nlp(text)

for token in doc:
    print(f"{token.text:<12} {token.dep_:<12} {token.head.text}")

html = displacy.render(doc, style="dep", jupyter=False)

display(HTML(html))

I            nsubj        saw
saw          ROOT         saw
the          det          man
man          dobj         saw
with         prep         saw
a            det          telescope
telescope    pobj         with
.            punct        saw


## Discuss with your partner:
## 1. What is the ROOT?
    saw 

## 2. What does "with a telescope" modify?
    spaCy attaches: with → saw
    So: with a telescope

## 3. Does the sentence have more than one meaning?
    Yes. The sentence is ambiguous.
    Possible meanings: I used a telescope to see the man.
                       The man had a telescope.

## 4. Did spaCy choose the interpretation you expected?
    spaCy chose:
        The speaker used the telescope.

## 5. Rewrite the sentence to remove ambiguity.
    Meaning 1:
        I used a telescope to see the man.

    Meaning 2:
        I saw the man who was carrying a telescope.

In [6]:
import spacy

nlp = spacy.load("en_core_web_sm")

text = """
Microsoft announced a new AI research initiative
in Berlin on 15 June 2026.

The project involves collaboration with Google,
SAP and SRH University Heidelberg.

Satya Nadella will present the initiative
at a conference in Germany.
"""

doc = nlp(text)

for ent in doc.ents:
    print(f"Entity Text   : {ent.text}")
    print(f"Entity Label  : {ent.label_}")
    print(f"Explanation   : {spacy.explain(ent.label_)}")
    print("-" * 40)

Entity Text   : Microsoft
Entity Label  : ORG
Explanation   : Companies, agencies, institutions, etc.
----------------------------------------
Entity Text   : AI
Entity Label  : ORG
Explanation   : Companies, agencies, institutions, etc.
----------------------------------------
Entity Text   : Berlin
Entity Label  : GPE
Explanation   : Countries, cities, states
----------------------------------------
Entity Text   : 15 June 2026
Entity Label  : DATE
Explanation   : Absolute or relative dates or periods
----------------------------------------
Entity Text   : Google
Entity Label  : ORG
Explanation   : Companies, agencies, institutions, etc.
----------------------------------------
Entity Text   : SAP
Entity Label  : ORG
Explanation   : Companies, agencies, institutions, etc.
----------------------------------------
Entity Text   : Satya Nadella
Entity Label  : PERSON
Explanation   : People, including fictional
----------------------------------------
Entity Text   : Germany
Entity Labe

# 1. Which entities were detected?
    Detected entities:

    Microsoft
    AI
    Berlin
    15 June 2026
    Google
    SAP
    Satya Nadella
    Germany

# 2. Which labels were assigned?
    | Entity        | Label  | Meaning      |
    | ------------- | ------ | ------------ |
    | Microsoft     | ORG    | Organization |
    | AI            | ORG    | Organization |
    | Berlin        | GPE    | Location     |
    | 15 June 2026  | DATE   | Date         |
    | Google        | ORG    | Organization |
    | SAP           | ORG    | Organization |
    | Satya Nadella | PERSON | Person       |
    | Germany       | GPE    | Location     |

# 3. Which entity types occur most often?
    Most frequent label: ORG
    Organizations detected:
        Microsoft
        AI
        Google
        SAP

# 4. Were any entities missed?
    Yes, possibly:
    “SRH University Heidelberg” was not detected
    “conference” was not detected as EVENT
    Also: AI was incorrectly classified as ORG.

In [7]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity

documents = [
    "AI is transforming healthcare",
    "Machine learning improves healthcare",
    "Football is a popular sport"
]

vectorizer = TfidfVectorizer()

tfidf_matrix = vectorizer.fit_transform(documents)

print("Vocabulary:")
print(vectorizer.get_feature_names_out())

print("\nTF-IDF Matrix:")
print(tfidf_matrix.toarray())

similarity_matrix = cosine_similarity(tfidf_matrix)

print("\nCosine Similarity Matrix:")
print(similarity_matrix)

print("Doc1 vs Doc2:", round(similarity_matrix[0][1], 3))
print("Doc1 vs Doc3:", round(similarity_matrix[0][2], 3))
print("Doc2 vs Doc3:", round(similarity_matrix[1][2], 3))


Vocabulary:
['ai' 'football' 'healthcare' 'improves' 'is' 'learning' 'machine'
 'popular' 'sport' 'transforming']

TF-IDF Matrix:
[[0.5628291  0.         0.42804604 0.         0.42804604 0.
  0.         0.         0.         0.5628291 ]
 [0.         0.         0.40204024 0.52863461 0.         0.52863461
  0.52863461 0.         0.         0.        ]
 [0.         0.52863461 0.         0.         0.40204024 0.
  0.         0.52863461 0.52863461 0.        ]]

Cosine Similarity Matrix:
[[1.         0.17209173 0.17209173]
 [0.17209173 1.         0.        ]
 [0.17209173 0.         1.        ]]
Doc1 vs Doc2: 0.172
Doc1 vs Doc3: 0.172
Doc2 vs Doc3: 0.0


# 1. Which pair has the highest similarity?
    Highest similarity:
        Doc1 vs Doc2 = 0.172
        Doc1 vs Doc3 = 0.172

# 2. Is this result reasonable?
    Partially yes.
    - Doc1 and Doc2 are related because both discuss healthcare.
    - Doc1 and Doc3 share only the word “is”.
    TF-IDF gives similarity based on shared words, not true meaning.

# 3. Why does TF-IDF treat some words as more important?
    TF-IDF gives higher importance to:
    - words that appear frequently in one document
    - but rarely across all documents
    Common words get lower importance.

# 4. Does cosine understand meaning?
    No.
    Cosine similarity only compares vector patterns and word frequencies.
    It does not truly understand semantics or context.

# 5. Would it know that "AI" and "machine learning" are related?
    No.
    TF-IDF treats them as completely different words unless they literally overlap.
    To understand semantic relationships, we need:
    - Word embeddings
    - Word2Vec
    - GloVe
    - BERT
    - Sentence Transformers
    These models capture meaning, unlike TF-IDF.